RecoTracker/TkTrackingRegions/plugins/L1MuonSeededTrackingRegionsProducer.h

```diff
@@ -0,0 +1,411 @@
+    edm::Handle<l1t::MuonBxCollection> muColl;
+    iEvent.getByToken(token_input, muColl);
+    if (muColl->size() == 0)
+      return result;
+
+    // always need the beam spot (as a fall back strategy for vertex modes)
+    edm::Handle<reco::BeamSpot> bs;
+    iEvent.getByToken(token_beamSpot, bs);
+    if (!bs.isValid())
+      return result;
+
+    // this is a default origin for all modes
+    GlobalPoint default_origin(bs->x0(), bs->y0(), bs->z0());
+
+    // vector of origin & halfLength pairs:
+    std::vector<std::pair<GlobalPoint, float>> origins;
+
+    // fill the origins and halfLengths depending on the mode
+    if (m_mode == BEAM_SPOT_FIXED || m_mode == BEAM_SPOT_SIGMA) {
+      origins.push_back(std::make_pair(
+          default_origin, (m_mode == BEAM_SPOT_FIXED) ? m_zErrorBeamSpot : m_nSigmaZBeamSpot * bs->sigmaZ()));
+    } else if (m_mode == VERTICES_FIXED || m_mode == VERTICES_SIGMA) {
+      edm::Handle<reco::VertexCollection> vertices;
+      iEvent.getByToken(token_vertex, vertices);
+      int n_vert = 0;
+      for (reco::VertexCollection::const_iterator v = vertices->begin();
+           v != vertices->end() && n_vert < m_maxNVertices;
+           ++v) {
+        if (v->isFake() || !v->isValid())
+          continue;
+
+        origins.push_back(std::make_pair(GlobalPoint(v->x(), v->y(), v->z()),
+                                         (m_mode == VERTICES_FIXED) ? m_zErrorVetex : m_nSigmaZVertex * v->zError()));
+        ++n_vert;
+      }
+      // no-vertex fall-back case:
+      if (origins.empty()) {
+        origins.push_back(std::make_pair(
+            default_origin, (m_nSigmaZBeamSpot > 0.) ? m_nSigmaZBeamSpot * bs->z0Error() : m_zErrorBeamSpot));
+      }
+    }
+
+    const MeasurementTrackerEvent* measurementTracker = nullptr;
+    if (!token_measurementTracker.isUninitialized()) {
+      edm::Handle<MeasurementTrackerEvent> hmte;
+      iEvent.getByToken(token_measurementTracker, hmte);
+      measurementTracker = hmte.product();
+    }
+
+    const auto& field = iSetup.getData(token_field);
+    const MultipleScatteringParametrisationMaker* msmaker = nullptr;
+    if (m_precise) {
+      msmaker = &iSetup.getData(token_msmaker);
+    }
+
+    // create tracking regions (maximum MaxNRegions of them) in directions of the
+    // objects of interest (we expect that the collection was sorted in decreasing pt order)
+    int n_regions = 0;
+    for (int ibx = muColl->getFirstBX(); ibx <= muColl->getLastBX() && n_regions < m_maxNRegions; ++ibx) {
+      if (centralBxOnly_ && (ibx != 0))
+        continue;
+
+      for (auto it = muColl->begin(ibx); it != muColl->end(ibx) && n_regions < m_maxNRegions; it++) {
+        unsigned int quality = it->hwQual();
+        if (quality <= l1MinQuality_)
+          continue;
+
+        float pt = it->pt();
+        float eta = it->eta();
+        if (pt < l1MinPt_ || std::abs(eta) > l1MaxEta_)
+          continue;
+
+        float theta = 2 * atan(exp(-eta));
```

Comment:
isn't this very expensive in a double loop? (note that I don't have a good suggestion to overcome this as you need it also to construct the momentum vector below - given the [dataformat](https://cmssdt.cern.ch/dxr/CMSSW/source/DataFormats/L1Trigger/interface/Muon.h?from=MuonBxCollection&case=true#12) - , but maybe some of the reviewers have)